File: Abeyonix_Backend/app/schemas/product.py
from pydantic import BaseModel
from typing import List, Optional
from datetime import datetime


class ProductMediaResponse(BaseModel):
    id: int
    url: str
    media_type: str
    is_primary: bool

    class Config:
        from_attributes = True


class PricingResponse(BaseModel):
    price: float
    discount_price: Optional[float]

    class Config:
        from_attributes = True



class InventoryResponse(BaseModel):
    quantity: int
    low_quantity_alert_at: int

    class Config:
        from_attributes = True



class ProductAttributeResponse(BaseModel):
    attribute_id: int
    attribute_name: str
    unit: Optional[str]
    value: str


class ProductDetailResponse(BaseModel):
    id: int
    name: str
    sku: str
    slug: str

    category_id: int
    category_name: Optional[str]

    sub_category_id: Optional[int]
    sub_category_name: Optional[str]

    brand: Optional[str]
    short_description: Optional[str]
    long_description: Optional[str]

    is_active: bool
    created_at: Optional[datetime]
    updated_at: Optional[datetime]

    pricing: PricingResponse
    inventory: InventoryResponse
    attributes: List[ProductAttributeResponse]
    media: List[ProductMediaResponse]

    class Config:
        from_attributes = True



# ---------------------------------------------------

class ProductListResponse(BaseModel):
    id: int
    name: str
    slug: str
    sku: str

    category_id: int
    category_name: Optional[str]

    sub_category_id: Optional[int]
    sub_category_name: Optional[str]

    brand: Optional[str]
    is_active: bool

    created_at: Optional[datetime]

    price: Optional[float]
    discount_price: Optional[float]

    class Config:
        from_attributes = True


class PaginatedProductResponse(BaseModel):
    total: int
    page: int
    page_size: int
    items: List[ProductDetailResponse]

File: Abeyonix_Frontend/src/types/product.ts
export interface ProductMedia {
  id: number;
  product_id: number;
  url: string;
  media_type: string;
  is_primary: boolean;
}

export interface ProductPricing {
  id: number;
  product_id: number;
  price: number;
  discount_price?: number | null;
  currency?: string;
}

export interface ProductInventory {
  id: number;
  product_id: number;
  quantity: number;
  low_quantity_alert_at?: number;
}

export interface ProductAttribute {
  attribute_id: number;
  attribute_name: string;
  unit?: string | null;
  value: string;
}

export interface ProductDetailResponse {
  id: number;
  name: string;
  sku: string;
  slug: string;

  category_id: number;
  category_name: string;

  sub_category_id?: number | null;
  sub_category_name?: string | null;

  brand?: string | null;
  short_description?: string | null;
  long_description?: string | null;

  is_active: boolean;
  created_at: string;
  updated_at: string;

  pricing: ProductPricing | null;
  inventory: ProductInventory | null;
  attributes: ProductAttribute[];
  media: ProductMedia[];
}

export interface PaginatedProductResponse {
  total: number;
  page: number;
  page_size: number;
  items: ProductDetailResponse[];
}



export interface ProductPayload {
  name: string;
  category_id: number;
  sub_category_id?: number;

  brand?: string;
  short_description?: string;
  long_description?: string;

  price: number;
  discount_price?: number;

  quantity: number;
  low_quantity_alert_at: number;

  attributes: {
    attribute_id: number;
    value: string;
  }[];

  images: File[];
  primary_image_index?: number;

  // update only
  sku?: string;
  existing_media_ids?: number[];
}
